cpp/src/arrow/compute/kernels/nth_to_indices.cc

```diff
@@ -0,0 +1,205 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/compute/kernels/nth_to_indices.h"
+
+#include <algorithm>
+#include <numeric>
+#include <vector>
+
+#include "arrow/builder.h"
+#include "arrow/compute/context.h"
+#include "arrow/compute/expression.h"
+#include "arrow/compute/logical_type.h"
+#include "arrow/type_traits.h"
+
+namespace arrow {
+
+class Array;
+
+namespace compute {
+
+/// \brief Kernel implementing NthToIndices operation
+class ARROW_EXPORT NthToIndicesKernel : public OpKernel {
+ protected:
+  std::shared_ptr<DataType> type_;
+
+ public:
+  /// \brief Kernel interface
+  ///
+  /// delegates to subclasses via NthToIndices()
+  virtual Status Call(FunctionContext* ctx, const Datum& values, uint64_t n,
+                      Datum* offsets) = 0;
```

Comment:
Is there a point in having virtual methods? Are you planning to create a second implementation of this kernel?

Reply:
removed virtual